x-pack/plugin/shutdown/src/internalClusterTest/java/org/elasticsearch/xpack/shutdown/AllocationFailuresResetOnShutdownIT.java

```diff
@@ -0,0 +1,302 @@
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        return Arrays.asList(MockIndexEventListener.TestPlugin.class, ShutdownPlugin.class);
+    }
+
+    public void testResetRelocationFailuresOnNodeShutdown() throws Exception {
+        String node1 = internalCluster().startNode();
+        createIndex("index1", 1, 0);
+        ensureGreen("index1");
+        final var failRelocation = new AtomicBoolean(true);
+        String node2 = internalCluster().startNode();
+        internalCluster().getInstances(MockIndexEventListener.TestEventListener.class)
+            .forEach(testEventListener -> testEventListener.setNewDelegate(new IndexEventListener() {
+                @Override
+                public void beforeIndexShardCreated(ShardRouting routing, Settings indexSettings) {
+                    if (failRelocation.get()) {
+                        throw new RuntimeException("FAIL");
+                    }
+                }
+            }));
+        updateIndexSettings(Settings.builder().put(INDEX_ROUTING_EXCLUDE_GROUP_PREFIX + "._name", node1), "index1");
+        ensureGreen("index1");
+        // await all relocation attempts are exhausted
+        var maxAttempts = MaxRetryAllocationDecider.SETTING_ALLOCATION_MAX_RETRY.get(Settings.EMPTY);
+        assertBusy(() -> {
+            var state = clusterAdmin().prepareState(TEST_REQUEST_TIMEOUT).get().getState();
+            var shard = state.routingTable().index("index1").shard(0).primaryShard();
+            assertThat(shard, notNullValue());
+            assertThat(shard.relocationFailureInfo().failedRelocations(), equalTo(maxAttempts));
+        });
+        // ensure the shard remain started
+        var stateAfterFailures = clusterAdmin().prepareState(TEST_REQUEST_TIMEOUT).get().getState();
+        var shardAfterFailures = stateAfterFailures.routingTable().index("index1").shard(0).primaryShard();
+        assertThat(shardAfterFailures, notNullValue());
+        assertThat(shardAfterFailures.state(), equalTo(ShardRoutingState.STARTED));
+        assertThat(stateAfterFailures.nodes().get(shardAfterFailures.currentNodeId()).getName(), equalTo(node1));
+        failRelocation.set(false);
+        // A non-RESTART node shutdown should reset the counter and allow more relocation retries
+        final var request = createRequest(randomFrom(SingleNodeShutdownMetadata.Type.values()), node1, node2);
+        client().execute(PutShutdownNodeAction.INSTANCE, request).get();
+        if (request.getType() != SingleNodeShutdownMetadata.Type.RESTART) {
+            assertBusy(() -> {
+                var stateAfterNodeJoin = internalCluster().clusterService().state();
+                var relocatedShard = stateAfterNodeJoin.routingTable().index("index1").shard(0).primaryShard();
+                assertThat(relocatedShard.relocationFailureInfo().failedRelocations(), Matchers.lessThan(maxAttempts));
```

Comment:
Probably for my knowledge: Why does this assert `failedRelocations < maxAttemps` instead of `failedRelocations == 0`?

Reply:
what we care about is the reset of the counter. so anything below max would do. I don't see a reason to be too strict here, since the test will be extra brittle if for whatever reason it takes more than one attempt to allocate/relocate.